module/zfs/ddt.c

```diff
@@ -575,9 +596,56 @@ ddt_remove(ddt_t *ddt, ddt_entry_t *dde)
 	ddt_free(dde);
 }
 
+/*
+ * Check if the DDT is over its quota.  This can be due to a few conditions:
+ *   1. 'dedup_table_quota' property is not 0 (none) and the ddt_get_ddt_dsize
+ *       exceeds this limit
+ *
+ *   2. 'dedup_table_quota' property is set to automatic and
+ *      a. the dedup allocation class could not satisfy a DDT allocation
+ *         in a recent transaction
+ *      b. the dedup allocation class has exceeded its calculated limit
+ */
+static boolean_t
+ddt_over_quota(spa_t *spa)
+{
+	if (spa->spa_dedup_table_quota == 0)
+		return (B_FALSE);
+
+	if (spa->spa_dedup_table_quota != UINT64_MAX)
+		return (ddt_get_ddt_dsize(spa) > spa->spa_dedup_table_quota);
+
+	/*
+	 * For automatic quota, table size is limited by the size of dedup class
+	 */
```

Comment:
For automatic quota's we also need to handle the case where the pool has a special allocation class but no dedup allocation class.  This way a pool with just a dedicated special device will still be able to write its DDT to the special device when auto is set (see `spa_preferred_class()`).  Applying the same 85% full limit seems reasonable.

Reply:
updated to also include special class